### PR TITLE
Implement upload()

### PR DESCRIPTION
**Pull request by JohanLorenzo**
~~WIP~~

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -1,2 +1,5 @@
 boto3
 taskcluster
+scriptworker
+-e ../../mozilla-releng/mozapkpublisher
+-e ../../mozilla-releng/beetmoverscript
```
Comment (JohanLorenzo) [inline requirements.txt]: TODO: Release new version of mozapkpublisher and beetmoverscript on Pypi

File: devedition_candidates.yml [hunk 2]
```diff
@@ -0,0 +1,15 @@
+---
+metadata:
+    name: "Beet Mover Manifest"
+    description: "Maps {{ product }} artifacts to pretty names for {{locale}} L10n repack"
```
Comment (lundjordan) [inline devedition_candidates.yml]: does this map en-US (non repacks) too?

File: staging_upload.py [hunk 3]
```diff
@@ -47,158 +65,152 @@ def die(msg):
...
+        s3_key='pub/devedition/candidates/{}-candidates/build{}/SHA512SUMS'.format(VERSION, BUILD_NUMBER),
+        path=general_shasum_file
+    )
+
+
+def _generate_sha512sums_file(context, path_with_sha512sums_per_platform):
+    # TODO this logic should not duplicate the one in move_beets()
+    lines = [
+        '{sha512}  update/{platform}/{locale}/devedition-{version}.complete.mar'.format(
+            sha512=details['sha512'], platform=platform, locale=locale, version=VERSION,
```
Comment (lundjordan) [inline staging_upload.py]: hm, shouldn't `details` be out of scope?

File: staging_upload.py [hunk 4]
```diff
@@ -5,31 +5,49 @@
...
+from beetmoverscript.script import upload_to_s3, move_beets, generate_beetmover_manifest
+from scriptworker.utils import download_file, raise_future_exceptions
 
 log = logging.getLogger(__name__)
-task_graph_id = "d996zZnrTJKne2LHXUTP6w"
-signing_task_names = ["signing-linux-devedition-nightly/opt"]
-num_expected_signing_tasks = 1
 max_concurrent_aiohttp_streams = 10
-# map to copy artifacts to bucket - public/build/target.complete.mar
+BUILD_NUMBER = 1    # XXX Edit this value
```
Comment (lundjordan) [inline staging_upload.py]: all these, mar locations, and locales list would be great if it came from a release specific config. not a blocker though

File: staging_upload.py [hunk 5]
```diff
@@ -5,31 +5,49 @@
...
+  'eo', 'es-AR', 'es-CL', 'es-ES', 'es-MX', 'et', 'eu', 'fa', 'ff', 'fi', 'fr',
+  'fy-NL', 'ga-IE', 'gd', 'gl', 'gn', 'gu-IN', 'he', 'hi-IN', 'hr', 'hsb', 'hu',
+  'hy-AM', 'id', 'is', 'it', 'ja-JP-mac', 'ka', 'kab', 'kk', 'km', 'kn', 'ko',
+  'lij', 'lt', 'lv', 'mai', 'mk', 'ml', 'mr', 'ms', 'nb-NO', 'nl', 'nn-NO', 'or',
+  'pa-IN', 'pl', 'pt-BR', 'pt-PT', 'rm', 'ro', 'ru', 'si', 'sk', 'sl', 'son',
+  'sq', 'sr', 'sv-SE', 'ta', 'te', 'th', 'tr', 'uk', 'ur', 'uz', 'vi', 'xh',
+  'zh-CN', 'zh-TW',
+)
+
 def get_filtered(graph):
```
Comment (lundjordan) [inline staging_upload.py]: is this used still?

**Comment by lundjordan** [comment]
overall looks good. I have some questions above. 

I'm also not sure about the output (bucket) results. copied from irc:

jlund> Jordan Lund jlorenzo: I'm trying to see why some platforms have more locales than others in your bucket upload?
12:36:45 e.g. only 6 locales here (no en-US) https://console.aws.amazon.com/s3/buckets/mozilla-releng-beet-mover-dev/pub/devedition/candidates/54.0b1-candidates/build1/update/linux-i686/?region=us-east-1
12:36:59 also, SHASUMS seems to be missing

File: staging_upload.py (hunk 3 above)
Comment (JohanLorenzo) [reply staging_upload.py]: Nope, `details` is defined in the list comprehension. I don't mind changing the name if it's confusing

File: devedition_candidates.yml (hunk 2 above)
Comment (JohanLorenzo) [reply devedition_candidates.yml]: Yes. I changed the description to avoid confusion. Thanks!

File: staging_upload.py [hunk 6]
```diff
@@ -5,31 +5,54 @@
...
-max_concurrent_aiohttp_streams = 10
-# map to copy artifacts to bucket - public/build/target.complete.mar
+# XXX There are less than 100 locales to upload. Having fewer connection open leads to about 5 timeouts. Here's why:
+# On S3, you first make a request to create an artifact. Then you upload it in a second request. Due to the number of
+# files to upload, some files time out before their upload start.
+# Another solution is to increase the timeout time. But if you have a good network connection (tested with an upload
+# speed of 25 MB/s), it's just faster to have more connections.
+max_concurrent_aiohttp_streams = 100
+BUILD_NUMBER = 1    # XXX Edit this value
+VERSION = '54.0b1'  # XXX Edit this value
```
Comment (JohanLorenzo) [inline staging_upload.py]: TODO: Change the version number before uploading and once we know it.

**Comment by escapewindow** [comment]
We've hit problems uploading more than 20 artifacts at the same time to s3... I think they throttle us.  I would be happy to be wrong.

**Comment by JohanLorenzo** [comment]
Oh :( I'll tweak the script to process batches of 10, then

File: staging_upload.py (hunk 3 above)
Comment (lundjordan) [reply staging_upload.py]: sorry, PEBK

I forgot to look down! :)

**Review by lundjordan — APPROVED** [review]
lgtm ⛵️ it

thanks for the context into your context :)

File: staging_upload.py (hunk 3 above)
Comment (lundjordan) [reply staging_upload.py]: PEBKAC, bah

**Comment by JohanLorenzo** [comment]
Uploads are now done by tens. ~~This is being tested at~~ 93 locales per platform, we're good to go! https://console.aws.amazon.com/s3/buckets/mozilla-releng-beet-mover-dev/pub/devedition/candidates/54.0b1-candidates/build2

**Comment by JohanLorenzo** [comment]
I don't remember what this PR is for 😅. Since it's more than 2.5 years old, I think it's safe to close it. We've had tremendous work done for staging releases since then.

**Pull request closed without merging**